mlflow/transformers/__init__.py

```diff
@@ -459,10 +469,12 @@ def save_model(
 
     code_dir_subpath = _validate_and_copy_code_paths(code_paths, str(path))
 
-    resolved_task = _get_or_infer_task_type(transformers_model, task)
+    resolved_transformers_task = _get_or_infer_transformers_task_type(transformers_model, task)
```

Comment:
Imo this method is a bit hard to follow, mainly because the `task` remains ambiguous i.e. whether it is transformers or inference task. Some method implies it handles transformer task such as `_validate_transformers_task_type`, but it actually inputs inference task as well.

To make things simple, I think we should rename it to `transformer_task` or `inference_task` as early as possible, and handle them in separate method, instead of having single method input `task`.

For example:
```
transformers_task, inference_task = _get_or_infer_task_type(transformers_model, task)

_validate_transformers_task_type(transformers_task)
_validate_inference_task_type(inference_task)

...

def _get_or_infer_task_type(transformers_model, task):
    if not task:
        transformers_task = _infer_transformers_task(task)
        inference_task = _get_inference_task(task, model)
    elif _is_transformer_task(task, ...):
        transformers_task = task
        inference_task = _get_inference_task(task, model)
    else:
        transformers_task = _infer_transformers_task(model)
        inference_task = task

    return transformers_task, inferece_task
```

This makes the downstream logic simple as well, like we can do `if inference_task` rather than checking supported types for pipeline. WDYT?

Reply:
I love that this makes the logic a lot simpler.
As a side note, do we want the `predict` function to always output a dictionary instead of a string even when `task` is not specified in the format of `llm/v1/*` and the input is in the normal Hugging Face format?

(Currently I haven't added the logic to infer the `inference_task` when it isn't provided as `task` yet.)

- If yes, it would change the behavior of `predict` quite a bit, but the implementation would be more straightforward
- If not, in addition to `inference_task`, we might also need to save whether `inference_task` is provided by user or inferred by us.

Either way, I will definitely need to refactor this a bit, but would love to hear your opinion of this.

Reply:
Hmm actually I'm not sure if we can "infer" the inference task type? We don't know if user intend to use it as chat or completion. If that's not deterministic, we can't output dictionary but just follow original transformers output string. This effectively means that user has to set `llm/v1/*` task if they want the new interface.

In other words, the condition here changes from
> Behavior change when (a) task is defined as llm/v1/completions or llm/v1/chat, or
(b) data, params are in OpenAI format

to 

> Behavior change when (a) task is defined as llm/v1/completions or llm/v1/chat, ##and###
(b) data, params are in OpenAI format

I think this is fine, but if not, we might need some fundamental change in the design....

CC: @dbczumar

Reply:
Got it! I think it would make sense to enforce it only when `task` is explicitly set as `llm/v1/*` -- it would be weird if people could pass arguments in OpenAI format but don't specify the `task` indeed.